data/src/history/manager.rs

```diff
@@ -439,13 +439,53 @@ impl Data {
 
         let filtered = messages
             .iter()
-            .filter(|message| {
+            .enumerate()
+            .filter(|(message_index, message)| {
                 if let message::Source::Server(Some(source)) = message.target.source() {
-                    !server_messages.get(source).exclude
+                    match server_messages.get(source).exclude {
+                        Exclude::All => false,
+                        Exclude::None => true,
+                        Exclude::Smart(seconds) => 'search_previous_messages: {
+                            for previous_message in
+                                messages.iter().rev().skip(messages.len() - *message_index)
+                            {
```

Comment:
Complexity grows here with the number of messages we need to iterate over.

Maybe a more stateful approach using Iter::scan to store the last message time for each user in a map would reduce that complexity. Though that has its own memory cost 🤔

Reply:
Looks good!